Initial implementation of branch prediction using a BTB

danbone opened the pull request:
Attempting to add some branch prediction logic to Olympia. I started with a branch target buffer.

Plan is to add a simple predictor, and a return address stack next. As this will require some form of recovery following speculative updates. Forming a foundation for further branch prediction modelling.

Early branch misprediction recovery could be done at a later date, but will require some more complex mechanism for restoring rename map before restarting.

Few points of interest:

- Adding flushing on misprediction has pushed the runtime up considerable, dhrystone takes 250s+ to simulate compared to 12secs on master. It's only doing 10k flushes, I'm surprised that it has had such an impact.
- The BTB model is basically just a clone of SimpleCache2 but the address decoder was a protected const, and I needed to change it. Any suggestion on a better way to do this?
- I intend to send all the flushed instructions to rename serially so to free up the renaming registers. This has an added bonus as it can be used to cancel the scoreboard callbacks. Unless there's a better suggestion?
- Should I update the JSON instruction generator? Will require new taken/target fields
- I want the branch prediction stuff encapsulated a bit more, I was thinking of having a new flush field in the instruction set on prediction, that can trigger flushes within the different modules. i.e. BTB miss could set flush to DECODE_FLUSH, branch misprediction sets EXECUTE_FLUSH. Would have to consider whether the instruction that triggers the flush is included or excluded from the flush. Thoughts?
- Any ideas on what tests to add?

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+            __LAST
+        };
+
+        BTBEntry() :
+            valid_(false)
+        { };
+
+        BTBEntry(const BTBEntry &rhs) :
+            BasicCacheItem(rhs),
+            valid_(rhs.valid_),
+            target_(rhs.target_),
+            is_call_(rhs.is_call_),
+            branchtype_(rhs.branchtype_),
+            lhist_counter_(rhs.lhist_counter_)
+        { };
+
+        BTBEntry &operator=(const BTBEntry &rhs)
+        {
+            if (&rhs != this) {
+                BasicCacheItem::operator=(rhs);
+                valid_ = rhs.valid_;
+                target_ = rhs.target_;
+                is_call_ = rhs.is_call_;
+                branchtype_ = rhs.branchtype_;
+                lhist_counter_ = rhs.lhist_counter_;
+            }
+            return *this;
+        }
```
You can do this:
```
```suggestion
        BTBEntry() = default;
        BTBEntry(const BTBEntry &rhs) = default;
        BTBEntry &operator=(const BTBEntry &rhs) = default;
```

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+            return *this;
+        }
+
+        void reset(uint64_t addr)
+        {
+            setValid(true);
+            setAddr(addr);
+            resetLHistCounter_();
+        }
```
I've never been a fan of `reset` methods.  Instead, you can do this after adding a new constructor that accepts an address:
```
   btb_entries_[idx] = BTBEntry(addr);
```

danbone replied:
I'm not sure I can do this (unless I'm not understanding something, fairly new to C++).

The problem is that we're using the [Cache ](https://github.com/sparcians/map/blob/master/sparta/cache/Cache.hpp)and [BasicCacheSet](https://github.com/sparcians/map/blob/master/sparta/cache/BasicCacheSet.hpp) as the underlying data structures. The BTB acts as a shim.

The Cache returns pointers to the BTBEntry objects stored within a vector inside BasicCacheSet, the user can then manipulate the object using the pointer, but there isn't a way to put a new object into the vector down in the Set class.

Is there a way I can use the constructor you suggested to copy into the pointer object on the RHS?

(I tried to experiment a bit, but couldn't get the syntax right - complains that the btb_entry is a pointer and if use new then I think you'd get a memory leak?)

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+            index_mask_ = (entries/num_ways)-1;
+            index_shift_ = sparta::utils::floor_log2(stride);
+        }
+        virtual uint64_t calcTag(uint64_t addr) const { return addr; }
+        virtual uint32_t calcIdx(uint64_t addr) const { return (addr >> index_shift_) & index_mask_; }
+        virtual uint64_t calcBlockAddr(uint64_t addr) const { return addr; }
+        virtual uint64_t calcBlockOffset(uint64_t addr) const { return 0;}
```
Are these new `virtual` methods or overrides from `AddrDecoderIF`?  If overrides, remove the `virtual` keyword and add `override` to the end.

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+        virtual uint64_t calcBlockAddr(uint64_t addr) const { return addr; }
+        virtual uint64_t calcBlockOffset(uint64_t addr) const { return 0;}
+    private:
+        uint32_t index_mask_;
+        uint32_t index_shift_;
```
Make these `const`

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+    public:
+        BTBAddrDecoder(uint32_t entries, uint32_t stride, uint32_t num_ways)
+        {
+            index_mask_ = (entries/num_ways)-1;
+            index_shift_ = sparta::utils::floor_log2(stride);
```
Move these to the initializer list

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+        /**
+         *\return Pointer to line with addr.  nullptr is returned if not fould
+        */
+        BTBEntry * getLine(uint64_t addr)
+        {
+            return cache_->getItem(addr);
+        }
```
Shouldn't this be a `const` method returning a `const` entry?

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+
+        // Allocate 'line' as having the new 'addr'
+        // 'line' doesn't know anything about NT
+        void allocateWithMRUUpdate(BTBEntry &line,
+                                    uint64_t   addr)
+        {
+            line.reset( addr );
+            touchMRU( line );
+        }
```
Why does this take a `line` and then reset it?  How about it creates one and returns it?

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+        }
+
+
+        void invalidateLineWithLRUUpdate(BTBEntry &line)
+        {
+            static const uint64_t addr=0;
+            line.reset( addr );
+            line.setValid( false );
+            touchLRU( line );
+        }
```
Same question here as above.

klingaard commented on core/BTB.hpp:
```diff
@@ -0,0 +1,277 @@
+        /**
+         * determine if there are any open ways in the set.
+         */
+        bool hasOpenWay(const uint64_t addr)
```
Can this method be made `const`?

klingaard commented on core/Fetch.cpp:
```diff
@@ -48,9 +50,49 @@ namespace olympia
     }
 
+
+    bool Fetch::predictInstruction_(InstPtr inst)
```
`const InstPtr & inst` to avoid the copy

klingaard commented on core/Fetch.cpp:
```diff
@@ -48,9 +50,49 @@ namespace olympia
+    {
+        if (btb_->isHit(inst->getPC()))
+        {
+            btb_hits_++;
```
Use pre-increment to avoid the copy (although the compiler will do that for you).

klingaard commented on core/BTB.hpp:
For the API, I suggest you don't use references as the parameters, but instead use pointers:
```
(const BTBEntry * line)
```
Reason for this, look at this example of buggy code:
```
    auto btb_entry = btb_->getEntry();
    doSomeStuffToUpdateEntry_(&btb_entry);
```
Does `getEntry` return reference or a copy?  The `auto` might make a copy and the call to `doSomeStuffToUpdateEntry_` will not update the original.

klingaard commented on core/Rename.cpp:
So the changes here reset the entirety of rename, preventing support for partial flushes.  :thinking: -- would you be up to updating the flushing criteria to include those instructions being flushed?

klingaard replied:
> Early branch misprediction recovery could be done at a later date, but will require some more complex mechanism for restoring rename map before restarting.

Ah, you already mentioned this...

arupc commented:
Thank you for initiating this. 

To provide more history, per previous discussion, we intend to add branch prediction support via an API, along with an implementation of simple branch predictor. This would enable use of a simple predictor while allowing for implementation of other predictors. There's also an interest in re-using existing branch predictor implementations

Please see  https://github.com/riscv-software-src/riscv-perf-model/issues/1 

I will start a discussion item on this topic. Let's collaborate via the discussion thread and review a proposal for branch prediction API

klingaard commented:
From our discussion in the SIG yesterday:

* This is the first PR that actually hammers on the flushing mechanism (and actually found a bug). 
* While this doesn't use the proposed branch API from @arupc and team, it's not a show-stopper for moving forward with this PR
* The flushing/redirect occurs in Decode, but only for a BTB miss (never seen before branch).  When this occurs the _entire pipe_ is flushed.  Not sure how safe this is right.  Might behoove us to actually implement a proper redirect from execution instead.
* The PR is missing some fundamental testing.  Specifically, we'd like to see a unit test _just for the BTB_ as it stands so it can be grown over time.  Also, augmentation of the JSON format might be handy to support taken/not taken branch behavior to test flushing, etc.

danbone commented:
- Updated branch with a new FlushingCriteria definition for determining which instructions to flush (inclusive/exclusive flush)
- Remove code which set scoreboard bits on flush which then led to fix flushing instructions from the issue queues in the executepipe/lsu that have outstanding scoreboard callbacks
- Updated rename map recovery to use iterate over the flushed instructions from the ROB and fix up the map/free registers.
- Moved ROB allocation from dispatch to rename. Ran into a race condition where instructions updated the rename map, but weren't included in the flush because they were stalled in dispatch and therefore hadn't been appended to the reorder buffer.

danbone commented:
Moving forward following the discussion in our last meeting on the 20th of December. I think we should separate out the changes needed for flushing. Then rebase the branch prediction API branch on top of it.
